http-response/background.js

```diff
@@ -1,22 +1,154 @@
 function listener(details) {
+  // The received data is a stream of bytes. In order to do text-based
+  // modifications, it is necessary to decode the bytes into a string
+  // using the proper character encoding, do any modifications, then
+  // encode back into a stream of bytes.
+  // Historically, detecting character encoding has been a tricky task
+  // taken on by the browser. Here, a simplified approach is taken
+  // and the complexity is hidden in a helper method.
+  let decoder, encoder;
+  [decoder, encoder] = detectCharsetAndSetupDecoderEncoder(details);
   let filter = browser.webRequest.filterResponseData(details.requestId);
-  let decoder = new TextDecoder("utf-8");
-  let encoder = new TextEncoder();
+  let fullStr = '';
+  
+  filter.ondata = e => {
+    // Note that the event's data may break in the middle of an encoded
+    // character - the stream parameter is critical for success as this
+    // method gets called multiple times.
+    let str = decoder.decode(e.data, {stream: true});
+    fullStr += str;

```

Comment:
Buffering everything before flushing is simple, but not user friendly.

Instead of buffering everything at once, you can consider flushing all but some penultimate bytes/characters. After all, if the string to replace is at most 7 characters, then you can safely flush everything except the last 6 characters, to handle this case:

```
[prefix this can be flushed]Exampl
```

Reply:
That's a good point. This is an artifact of the code I was porting this from (where I needed to do a regex search of unknown length) and can be improved. I think the best solution for this would be a streaming regex transformer - I don't suppose something along those lines exists without external library support? That would make my day.  Otherwise, I will change it as requested.

Reply:
You don't need an external library for this, the logic can be written in a concise way. See the brief explanation below for the algorithm. It is definitely more involved than buffering all and flushing at once though. If the example becomes very unwieldy if this were to be implemented, then buffering may not be that bad of an idea.

Instead of using `replace` with a regexp and a string, you can use the `.exec` method of a regular expression with the `g` flag to find matches. Flush the prefix, append the replacement, and continue (if you are cutting the string, reset `lastIndex` to `0` to resume from the start, if you aren't cutting the string you can just call `.exec` again and it will continue searching from position `lastIndex` in the string).

When you don't find a match, and know that the regular expression has a maximum number of matches (such as `8` in case of `/Example|Test/g`), then cut the remaining string in two parts: Keep the last 7 characters (=max max length minus one) for the next time and flush the rest (this is what I explained in my previous comment).

Note: when your regular expression does not have a unique match (e.g. `/A{1,3}/g` or `[0-9]+`), then you need extra logic: to make sure that as much is matched as possible, when the regular expression matches the end of the string collected so far, you should not flush and keep the string for the next iteration, in case the regexp is going to match more.

Reply:
I opted to go for the simpler static string streaming replacement, and added comments around its limits.